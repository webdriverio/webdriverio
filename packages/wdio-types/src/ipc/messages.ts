import type { Options, Capabilities } from '@wdio/types'
import type { BrowserData, SnapshotResultMessage, SessionEndedMessage } from '../Runner.js'

export enum IPC_MESSAGE_TYPES {
    snapshotResultMessage = 100,
    sessionEnded,
    sessionStartedMessage,
    readyEventMessage,
    printFailureMessage,
    reporterRealTime,
    errorMessage,
    testFrameworkInitMessage,
    finishedCommandMessage,
    debuggerMessage,
    sessionMetadataMessage
}

export type IPCMessageValue = {
    [IPC_MESSAGE_TYPES.snapshotResultMessage]: SnapshotResultMessage
    [IPC_MESSAGE_TYPES.sessionEnded]: SessionEndedMessage
    [IPC_MESSAGE_TYPES.readyEventMessage]: ReadyEvent
    [IPC_MESSAGE_TYPES.printFailureMessage]: PrintFailureMessagePayload
    [IPC_MESSAGE_TYPES.reporterRealTime]: ReporterRealTime
    [IPC_MESSAGE_TYPES.errorMessage]: Error
    [IPC_MESSAGE_TYPES.testFrameworkInitMessage]: TestFrameworkInit,
    [IPC_MESSAGE_TYPES.sessionStartedMessage]: SessionStartedMessagePayload
    [IPC_MESSAGE_TYPES.finishedCommandMessage]: FinishedCommand
    [IPC_MESSAGE_TYPES.debuggerMessage]: Debugger
    [IPC_MESSAGE_TYPES.sessionMetadataMessage]: SessionMetadata
}

export type IPCMessage<T extends IPC_MESSAGE_TYPES> = {
    type: T
    value: IPCMessageValue[T]
}

export type AnyIPCMessage = IPCMessage<IPC_MESSAGE_TYPES>

interface TestFrameworkInit{
    cid: string
    caps: Capabilities.RequestedStandaloneCapabilities
    specs: string[]
    hasTests: boolean
}

interface Error {
    origin: string;
    name: string;
    content: {
        name?: string,
        message?: string,
        stack?: string,
        cid?: string,
        error?: {
            message: string
        },
        fullTitle?: string,
        type?: string,
        state?: string
    }
}

interface PrintFailureMessagePayload {
    cid?: string
    specs?: string[]
    uid?: string
    file?: string
    title?: string
    error?: string
    sessionId?: string
    config?: unknown
    isMultiremote?: boolean
    instanceOptions?: Options.Testrunner
    capabilities?: unknown
    retry?: number
}

export interface SessionMetadata {
    automationProtocol?: string
    sessionId: string
    isW3C: boolean
    protocol?: string
    hostname?: string
    port?: number
    path?: string
    queryParams?: Record<string, string>
    isMultiremote: boolean
    instances?: Record<string, Partial<BrowserData>>
    capabilities: WebdriverIO.Capabilities
    injectGlobals?: boolean
    headers?: Record<string, string>
}

export interface SessionStartedMessagePayload {
    origin: 'worker'
    name: 'sessionStarted'
    content: {
        sessionId: string
        isW3C: boolean
        protocol: string
        hostname: string
        port: number
        path: string
        headers: Record<string, string>
        isMultiremote: boolean
        injectGlobals: boolean
        capabilities: WebdriverIO.Capabilities
    },
    cid?: string
}

interface ReadyEvent {
    name: string,
    origin: string
}

interface FinishedCommand {
    origin: string,
    name: string,
    content: {
        command: string,
        result: unknown
    }
}

type Debugger =
    | { origin: 'debugger'; name: 'start'; content?: never }
    | { origin: 'debugger'; name: 'eval'; content: { cmd: string } }
    | { origin: 'debugger'; name: 'stop'; content?: never }

interface ReporterRealTime { origin: string, name: string, content: string }
