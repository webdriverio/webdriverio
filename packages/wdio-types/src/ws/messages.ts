import type { Event } from '../Runner.js'
import type { $, $$ } from '@wdio/globals'

export enum WS_MESSAGE_TYPES {
    consoleMessage = 200,
    commandRequestMessage,
    commandResponseMessage,
    hookTriggerMessage,
    hookResultMessage,
    expectRequestMessage,
    expectResponseMessage,
    expectMatchersRequest,
    expectMatchersResponse,
    coverageMap,
    customCommand,
    initiateBrowserStateRequest,
    initiateBrowserStateResponse,
    browserTestResult,
    workerResponseMessage
}

export type WSMessageValue = {
    [WS_MESSAGE_TYPES.consoleMessage]: ConsoleEvent
    [WS_MESSAGE_TYPES.commandRequestMessage]: CommandRequestEvent
    [WS_MESSAGE_TYPES.commandResponseMessage]: CommandResponseEvent
    [WS_MESSAGE_TYPES.hookTriggerMessage]: HookTriggerEvent
    [WS_MESSAGE_TYPES.hookResultMessage]: HookResultEvent
    [WS_MESSAGE_TYPES.expectRequestMessage]: ExpectRequestEvent
    [WS_MESSAGE_TYPES.expectResponseMessage]: ExpectResponseEvent
    [WS_MESSAGE_TYPES.expectMatchersRequest]: never
    [WS_MESSAGE_TYPES.expectMatchersResponse]: ExpectMatchersResponse
    [WS_MESSAGE_TYPES.coverageMap]: CoverageMapData
    [WS_MESSAGE_TYPES.customCommand]: CustomCommandEvent
    [WS_MESSAGE_TYPES.initiateBrowserStateRequest]: BrowserStateRequest
    [WS_MESSAGE_TYPES.initiateBrowserStateResponse]: BrowserState
    [WS_MESSAGE_TYPES.browserTestResult]: BrowserTestResults
    [WS_MESSAGE_TYPES.workerResponseMessage]: WorkerResponseMessage
}

export type WSMessage<T extends WS_MESSAGE_TYPES> = {
    type: T
    value: WSMessageValue[T]
}

export type AnyWSMessage = WSMessage<WS_MESSAGE_TYPES>

import type { CoverageMapData } from 'istanbul-lib-coverage'

// Placeholder type declarations â€“ replace with actual imports
interface ConsoleEvent {
    name: 'consoleEvent'
    type: 'log' | 'info' | 'warn' | 'debug' | 'error'
    args: unknown[]
    cid: string
}

interface CommandRequestEvent {
    id: number;
    cid: string;
    commandName: string;
    args: unknown[];
    scope?: string
}

interface CommandResponseEvent {
    id: number;
    result?: unknown;
    error?: Error
}

interface HookTriggerEvent {
    id: number;
    cid: string;
    name: string;
    args: unknown[]
}

interface HookResultEvent {
    id: number;
    error?: Error
}

interface ExpectRequestEvent {
    id: number
    cid: string
    matcherName: string
    scope: unknown
    args: unknown[]
    element?: ReturnType<typeof $> | ReturnType<typeof $$>[number] | undefined
    context?: 'WebdriverIO.Element' | 'WebdriverIO.Browser' | string
    errorStack?: string
}

interface ExpectResponseEvent {
    id: number;
    pass: boolean;
    message: string
}

interface ExpectMatchersResponse {
    matchers: string[]
}

interface CustomCommandEvent {
    commandName: string;
    cid: string
}

interface BrowserStateRequest {
    cid: string
}

interface BrowserState {
    customCommands: string[]
}
interface BrowserTestResults {
    failures: number
    events: Event[]
}

export interface WorkerResponseMessage {
    origin: string
    name: string
    args: {
        id: number
        message: AnyWSMessage
    }
}

