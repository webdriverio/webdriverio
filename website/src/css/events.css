.profileStylesNoPhoto {
    width: 30px;
    transform: translateY(10px);
    margin-right: 5px;
}

.profileStyles {
    /* ...profileStylesNoPhoto; */
    border-radius: 50%;
    border: 1px solid #EA5907;
}

.profileLinkStyles {
    background-color: transparent!important;
    border-bottom: none!important;
    padding-right: 15px!important;
}

.event {
    display: flex;
    align-items: flex-start;
    gap: 20px;
    margin-bottom: 20px;
}

.eventInput {
    padding: 10px;
    margin-bottom: 10px;
    width: 50%;
    border-radius: 3px;
}
.eventButton {
    padding: 10px;
    border-radius: 3px;
    background-color: var(--ifm-color-primary);
    color: white;
    border: none;
    font-weight: bold;
    cursor: pointer;
}

.footnote {
    font-size: 0.6em;
    width: 500px;
    padding: 10px 0;
}

@media (max-width: 1150px) {
    .event {
        display: block;
        margin-bottom: 50px;
    }

    .event > a > img {
        width: 100%;
    }
}
