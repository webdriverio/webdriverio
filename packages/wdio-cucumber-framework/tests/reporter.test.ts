import { EventEmitter } from 'events'
import { messages } from '@cucumber/messages'

import CucumberReporter from '../src/reporter'
import {
    gherkinDocument,
    pickle,
    testRunStarted,
    testCase,
    testCaseStarted,
    testStepStarted,
    testStepFinished,
    testCaseFinished,
    testRunFinished
} from './fixtures/envelopes'

const wdioReporter = {
    write: jest.fn(),
    emit: jest.fn(),
    on: jest.fn()
}

const buildGherkinDocEvent = () => gherkinDocument
const gherkinDocEvent = buildGherkinDocEvent()
const gherkinDocEventNoLine = buildGherkinDocEvent()
delete gherkinDocEventNoLine.feature?.location?.line

const loadGherkin = (eventBroadcaster: EventEmitter) =>
    eventBroadcaster.emit('envelope', { gherkinDocument })
const loadGherkinNoLine = (eventBroadcaster: EventEmitter) =>
    eventBroadcaster.emit('envelope', { gherkinDocument: gherkinDocEventNoLine })
const acceptPickle = (eventBroadcaster: EventEmitter) =>
    eventBroadcaster.emit('envelope', { pickle })
const prepareSuite = (eventBroadcaster: EventEmitter) =>
    eventBroadcaster.emit('envelope', { testCase })
const startSuite = (eventBroadcaster: EventEmitter) =>
    eventBroadcaster.emit('envelope', { testCaseStarted })

describe('cucumber reporter', () => {
    describe('emits messages for certain cucumber events', () => {
        const cid = '0-1'
        const specs = ['/foobar.js']
        let eventBroadcaster: EventEmitter
        let cucumberReporter: CucumberReporter

        beforeEach(() => {
            wdioReporter.emit.mockClear()
            eventBroadcaster = new EventEmitter()
            cucumberReporter = new CucumberReporter(
                eventBroadcaster,
                {} as any,
                { failAmbiguousDefinitions: true } as any,
                cid,
                specs,
                wdioReporter as any
            )
        })

        it('should not send any data on `gherkin-document` event', () => {
            loadGherkin(eventBroadcaster)
            expect(cucumberReporter.eventListener['_gherkinDocEvents']).toEqual([gherkinDocEvent])
            expect(wdioReporter.emit).not.toHaveBeenCalled()
        })

        it('should send proper data on `test-run-started` event', () => {
            loadGherkin(eventBroadcaster)
            wdioReporter.emit.mockClear()
            eventBroadcaster.emit('envelope', { testRunStarted: {} })
            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })

        it('should not send any data on `pickle-accepted` event', () => {
            loadGherkin(eventBroadcaster)
            wdioReporter.emit.mockClear()
            acceptPickle(eventBroadcaster)

            expect(wdioReporter.emit).not.toHaveBeenCalled()
        })

        it('should not be ok if line is missing', () => {
            loadGherkinNoLine(eventBroadcaster)
            wdioReporter.emit.mockClear()
            acceptPickle(eventBroadcaster)

            expect(wdioReporter.emit).not.toHaveBeenCalled()
        })

        it('should send accepted pickle\'s data on `test-case-started` event', () => {
            loadGherkin(eventBroadcaster)
            acceptPickle(eventBroadcaster)
            prepareSuite(eventBroadcaster)
            wdioReporter.emit.mockClear()
            startSuite(eventBroadcaster)
            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })

        describe('step finished events', () => {
            beforeEach(() => {
                loadGherkin(eventBroadcaster)
                acceptPickle(eventBroadcaster)
                prepareSuite(eventBroadcaster)
                startSuite(eventBroadcaster)
                wdioReporter.emit.mockClear()
            })

            it('should send proper data on `test-step-started` event', () => {
                eventBroadcaster.emit('envelope', { testStepStarted })
                expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
            })

            it('passed step', () => {
                const passingStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))
                eventBroadcaster.emit('envelope', { testStepFinished: passingStep })
                delete wdioReporter.emit.mock.calls[0][1].duration
                expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
            })

            it('failed step', () => {
                const failedStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))
                failedStep.testStepResult = {
                    ...failedStep.testStepResult,
                    status: 6
                }
                eventBroadcaster.emit('envelope', { testStepFinished: failedStep })
                delete wdioReporter.emit.mock.calls[0][1].duration
                expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
            })
        })

        it('should send proper data on onTestRunFinished', () => {
            const passingStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))

            loadGherkin(eventBroadcaster)
            acceptPickle(eventBroadcaster)
            prepareSuite(eventBroadcaster)
            startSuite(eventBroadcaster)
            eventBroadcaster.emit('envelope', { testStepStarted })
            eventBroadcaster.emit('envelope', { testStepFinished: passingStep })
            eventBroadcaster.emit('envelope', { testCaseFinished })
            wdioReporter.emit.mockClear()

            eventBroadcaster.emit('envelope', { testRunFinished })
            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })

        it('should proper data when executing a hook', () => {
            const hookStarted: messages.ITestStepStarted = JSON.parse(JSON.stringify(testStepStarted))
            hookStarted.testStepId = '24'
            const hookFinished: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))
            hookFinished.testStepId = '24'

            loadGherkin(eventBroadcaster)
            acceptPickle(eventBroadcaster)
            prepareSuite(eventBroadcaster)
            startSuite(eventBroadcaster)
            eventBroadcaster.emit('envelope', { testStepStarted: hookStarted })
            wdioReporter.emit.mockClear()
            eventBroadcaster.emit('envelope', { testStepFinished: hookFinished })
            delete wdioReporter.emit.mock.calls[0][1].duration
            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })
    })

    describe('emits messages for certain cucumber events when executed in scenarioLeverReporter', () => {
        const cid = '0-1'
        const specs = ['/foobar.js']
        let eventBroadcaster: EventEmitter
        let cucumberReporter: CucumberReporter

        beforeEach(() => {
            eventBroadcaster = new EventEmitter()
            cucumberReporter = new CucumberReporter(eventBroadcaster, {} as any, { failAmbiguousDefinitions: true, scenarioLevelReporter: true } as any, cid, specs, wdioReporter as any)
        })

        it('should not send any data on `gherkin-document` event', () => {
            loadGherkin(eventBroadcaster)
            expect(cucumberReporter.eventListener['_gherkinDocEvents']).toEqual([gherkinDocEvent])
            expect(wdioReporter.emit).not.toHaveBeenCalled()
        })

        it('should send proper data on `test-run-started` event', () => {
            loadGherkin(eventBroadcaster)
            wdioReporter.emit.mockClear()
            eventBroadcaster.emit('envelope', { testRunStarted })
            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })

        it('should not send any data on `pickle-accepted` event', () => {
            loadGherkin(eventBroadcaster)
            wdioReporter.emit.mockClear()
            acceptPickle(eventBroadcaster)

            expect(wdioReporter.emit).not.toHaveBeenCalled()
        })

        it('should not be ok if line is missing', () => {
            loadGherkinNoLine(eventBroadcaster)
            wdioReporter.emit.mockClear()
            acceptPickle(eventBroadcaster)

            expect(wdioReporter.emit).not.toHaveBeenCalled()
        })

        it('should send accepted pickle\'s data on `test-case-started` event', () => {
            loadGherkin(eventBroadcaster)
            acceptPickle(eventBroadcaster)
            prepareSuite(eventBroadcaster)
            wdioReporter.emit.mockClear()
            startSuite(eventBroadcaster)

            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })

        it('should send proper data on `test-case-finished` event', () => {
            const passingStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))

            loadGherkin(eventBroadcaster)
            acceptPickle(eventBroadcaster)
            prepareSuite(eventBroadcaster)
            startSuite(eventBroadcaster)
            eventBroadcaster.emit('envelope', { testStepStarted })
            eventBroadcaster.emit('envelope', { testStepFinished: passingStep })
            wdioReporter.emit.mockClear()

            eventBroadcaster.emit('envelope', { testCaseFinished })
            delete wdioReporter.emit.mock.calls[0][1].duration
            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })

        it('should send proper data on onTestRunFinished', () => {
            const passingStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))

            loadGherkin(eventBroadcaster)
            acceptPickle(eventBroadcaster)
            prepareSuite(eventBroadcaster)
            startSuite(eventBroadcaster)
            eventBroadcaster.emit('envelope', { testStepStarted })
            eventBroadcaster.emit('envelope', { testStepFinished: passingStep })
            eventBroadcaster.emit('envelope', { testCaseFinished })
            wdioReporter.emit.mockClear()

            eventBroadcaster.emit('envelope', { testRunFinished })
            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })
    })

    describe('provides a fail counter', () => {
        let eventBroadcaster: EventEmitter
        let reporter: CucumberReporter

        beforeEach(() => {
            eventBroadcaster = new EventEmitter()
            reporter = new CucumberReporter(eventBroadcaster, {} as any, { failAmbiguousDefinitions: true } as any, '0-1', ['/foobar.js'], wdioReporter as any)

            loadGherkin(eventBroadcaster)
            acceptPickle(eventBroadcaster)
            prepareSuite(eventBroadcaster)
            startSuite(eventBroadcaster)
        })

        it('should increment failed counter on `failed` status', () => {
            const failedStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))
            failedStep.testStepResult = {
                ...failedStep.testStepResult,
                status: 6
            }
            eventBroadcaster.emit('envelope', { testStepFinished: failedStep })
            expect(reporter.failedCount).toBe(1)
        })

        it('should increment failed counter on `ambiguous` status', () => {
            const ambiguousStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))
            ambiguousStep.testStepResult = {
                ...ambiguousStep.testStepResult,
                status: 5
            }
            eventBroadcaster.emit('envelope', { testStepFinished: ambiguousStep })
            expect(reporter.failedCount).toBe(1)
        })

        it('should increment failed counter on `undefined` status', () => {
            const undefinedStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))
            undefinedStep.testStepResult = {
                ...undefinedStep.testStepResult,
                status: 4
            }
            eventBroadcaster.emit('envelope', { testStepFinished: undefinedStep })
            expect(reporter.failedCount).toBe(1)
        })

        it('should not increment failed counter on `undefined` status if ignoreUndefinedDefinitions set to true', () => {
            reporter['_options'].ignoreUndefinedDefinitions = true
            const undefinedStep: messages.ITestStepFinished = JSON.parse(JSON.stringify(testStepFinished))
            undefinedStep.testStepResult = {
                ...undefinedStep.testStepResult,
                status: 4
            }
            eventBroadcaster.emit('envelope', { testStepFinished: undefinedStep })
            expect(reporter.failedCount).toBe(0)
        })
    })

    describe('tags in title', () => {
        let eventBroadcaster: EventEmitter

        beforeAll(() => {
            eventBroadcaster = new EventEmitter()
            new CucumberReporter(eventBroadcaster, {} as any, {
                tagsInTitle: true
            } as any, '0-1', ['/foobar.js'], wdioReporter as any)
        })

        it('should add tags on handleBeforeFeatureEvent', () => {
            eventBroadcaster.emit('envelope', { gherkinDocument: gherkinDocEvent })
            expect(wdioReporter.emit).not.toBeCalled()
            eventBroadcaster.emit('envelope', { testRunStarted })

            expect(wdioReporter.emit.mock.calls).toMatchSnapshot()
        })
    })

    afterEach(() => {
        wdioReporter.on.mockClear()
        wdioReporter.write.mockClear()
        wdioReporter.emit.mockClear()
    })
})
