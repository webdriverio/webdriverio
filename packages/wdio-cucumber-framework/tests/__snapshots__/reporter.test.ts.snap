// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cucumber reporter emits messages for certain cucumber events should proper data when executing a hook 1`] = `
Array [
  Array [
    "hook:end",
    Object {
      "argument": undefined,
      "cid": "0-1",
      "error": undefined,
      "featureName": "Example feature",
      "file": "/some/path/to/features/my-feature.feature",
      "parent": "0",
      "scenarioName": "Get size of an element",
      "specs": Array [
        "/foobar.js",
      ],
      "state": "PASSED",
      "tags": Array [
        Object {
          "astNodeId": "9",
          "name": "@feature-tag1",
        },
        Object {
          "astNodeId": "10",
          "name": "@feature-tag2",
        },
        Object {
          "astNodeId": "5",
          "name": "@skip(browserName='chrome';platformName='windows')",
        },
        Object {
          "astNodeId": "6",
          "name": "@scenario-tag1",
        },
        Object {
          "astNodeId": "7",
          "name": "@scenario-tag2",
        },
      ],
      "title": "",
      "type": "hook",
      "uid": "24",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events should send accepted pickle's data on \`test-case-started\` event 1`] = `
Array [
  Array [
    "suite:start",
    Object {
      "cid": "0-1",
      "file": "/some/path/to/features/my-feature.feature",
      "fullTitle": "my-feature.feature:undefined:1: Get size of an element",
      "parent": "my-feature.feature:undefined:1",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "astNodeId": "9",
          "name": "@feature-tag1",
        },
        Object {
          "astNodeId": "10",
          "name": "@feature-tag2",
        },
        Object {
          "astNodeId": "5",
          "name": "@skip(browserName='chrome';platformName='windows')",
        },
        Object {
          "astNodeId": "6",
          "name": "@scenario-tag1",
        },
        Object {
          "astNodeId": "7",
          "name": "@scenario-tag2",
        },
      ],
      "title": "Get size of an element",
      "type": "scenario",
      "uid": "0",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events should send proper data on \`test-run-started\` event 1`] = `
Array [
  Array [
    "suite:start",
    Object {
      "cid": "0-1",
      "description": "  As a user of WebdriverIO
  I should be able to use different commands
  to get informations about elements on the page",
      "file": "/some/path/to/features/my-feature.feature",
      "keyword": "Feature",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "id": "9",
          "location": Object {
            "column": 1,
            "line": 1,
          },
          "name": "@feature-tag1",
        },
        Object {
          "id": "10",
          "location": Object {
            "column": 1,
            "line": 2,
          },
          "name": "@feature-tag2",
        },
      ],
      "title": "Example feature",
      "type": "feature",
      "uid": "my-feature.feature:undefined:1",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events should send proper data on onTestRunFinished 1`] = `
Array [
  Array [
    "suite:end",
    Object {
      "cid": "0-1",
      "duration": NaN,
      "file": "/some/path/to/features/my-feature.feature",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "id": "9",
          "location": Object {
            "column": 1,
            "line": 1,
          },
          "name": "@feature-tag1",
        },
        Object {
          "id": "10",
          "location": Object {
            "column": 1,
            "line": 2,
          },
          "name": "@feature-tag2",
        },
      ],
      "title": "Example feature",
      "type": "feature",
      "uid": "my-feature.feature:undefined:1",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events step finished events failed step 1`] = `
Array [
  Array [
    "test:fail",
    Object {
      "argument": Object {
        "rows": Array [
          Object {
            "cells": Array [
              "Item",
              "Amount",
            ],
          },
          Object {
            "cells": Array [
              "Milk",
              "2",
            ],
          },
          Object {
            "cells": Array [
              "Butter",
              "1",
            ],
          },
        ],
      },
      "cid": "0-1",
      "error": [Error],
      "featureName": "Example feature",
      "file": "/some/path/to/features/my-feature.feature",
      "fullTitle": "0: should the element '.header-logged-out a' be 32px wide and 35px high",
      "parent": "0",
      "passed": false,
      "scenarioName": "Get size of an element",
      "specs": Array [
        "/foobar.js",
      ],
      "state": "fail",
      "tags": Array [
        Object {
          "astNodeId": "9",
          "name": "@feature-tag1",
        },
        Object {
          "astNodeId": "10",
          "name": "@feature-tag2",
        },
        Object {
          "astNodeId": "5",
          "name": "@skip(browserName='chrome';platformName='windows')",
        },
        Object {
          "astNodeId": "6",
          "name": "@scenario-tag1",
        },
        Object {
          "astNodeId": "7",
          "name": "@scenario-tag2",
        },
      ],
      "title": "should the element '.header-logged-out a' be 32px wide and 35px high",
      "type": "step",
      "uid": "12",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events step finished events passed step 1`] = `
Array [
  Array [
    "test:undefined",
    Object {
      "argument": Object {
        "rows": Array [
          Object {
            "cells": Array [
              "Item",
              "Amount",
            ],
          },
          Object {
            "cells": Array [
              "Milk",
              "2",
            ],
          },
          Object {
            "cells": Array [
              "Butter",
              "1",
            ],
          },
        ],
      },
      "cid": "0-1",
      "error": undefined,
      "featureName": "Example feature",
      "file": "/some/path/to/features/my-feature.feature",
      "fullTitle": "0: should the element '.header-logged-out a' be 32px wide and 35px high",
      "parent": "0",
      "passed": false,
      "scenarioName": "Get size of an element",
      "specs": Array [
        "/foobar.js",
      ],
      "state": "undefined",
      "tags": Array [
        Object {
          "astNodeId": "9",
          "name": "@feature-tag1",
        },
        Object {
          "astNodeId": "10",
          "name": "@feature-tag2",
        },
        Object {
          "astNodeId": "5",
          "name": "@skip(browserName='chrome';platformName='windows')",
        },
        Object {
          "astNodeId": "6",
          "name": "@scenario-tag1",
        },
        Object {
          "astNodeId": "7",
          "name": "@scenario-tag2",
        },
      ],
      "title": "should the element '.header-logged-out a' be 32px wide and 35px high",
      "type": "step",
      "uid": "12",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events step finished events should send proper data on \`test-step-started\` event 1`] = `
Array [
  Array [
    "test:start",
    Object {
      "argument": Object {
        "rows": Array [
          Object {
            "cells": Array [
              "Item",
              "Amount",
            ],
          },
          Object {
            "cells": Array [
              "Milk",
              "2",
            ],
          },
          Object {
            "cells": Array [
              "Butter",
              "1",
            ],
          },
        ],
      },
      "cid": "0-1",
      "featureName": "Example feature",
      "file": "/some/path/to/features/my-feature.feature",
      "fullTitle": "0: Then should the element '.header-logged-out a' be 32px wide and 35px high",
      "parent": "0",
      "scenarioName": "Get size of an element",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "astNodeId": "9",
          "name": "@feature-tag1",
        },
        Object {
          "astNodeId": "10",
          "name": "@feature-tag2",
        },
        Object {
          "astNodeId": "5",
          "name": "@skip(browserName='chrome';platformName='windows')",
        },
        Object {
          "astNodeId": "6",
          "name": "@scenario-tag1",
        },
        Object {
          "astNodeId": "7",
          "name": "@scenario-tag2",
        },
      ],
      "title": "Then should the element '.header-logged-out a' be 32px wide and 35px high",
      "type": "test",
      "uid": "12",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events when executed in scenarioLeverReporter should send accepted pickle's data on \`test-case-started\` event 1`] = `
Array [
  Array [
    "test:start",
    Object {
      "cid": "0-1",
      "file": "/some/path/to/features/my-feature.feature",
      "fullTitle": "my-feature.feature:undefined:1: Get size of an element",
      "parent": "my-feature.feature:undefined:1",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "astNodeId": "9",
          "name": "@feature-tag1",
        },
        Object {
          "astNodeId": "10",
          "name": "@feature-tag2",
        },
        Object {
          "astNodeId": "5",
          "name": "@skip(browserName='chrome';platformName='windows')",
        },
        Object {
          "astNodeId": "6",
          "name": "@scenario-tag1",
        },
        Object {
          "astNodeId": "7",
          "name": "@scenario-tag2",
        },
      ],
      "title": "Get size of an element",
      "type": "scenario",
      "uid": "0",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events when executed in scenarioLeverReporter should send proper data on \`test-case-finished\` event 1`] = `
Array [
  Array [
    "test:undefined",
    Object {
      "argument": undefined,
      "cid": "0-1",
      "error": undefined,
      "featureName": "Example feature",
      "file": "/some/path/to/features/my-feature.feature",
      "parent": "0",
      "passed": false,
      "scenarioName": "Get size of an element",
      "specs": Array [
        "/foobar.js",
      ],
      "state": "undefined",
      "tags": Array [
        Object {
          "astNodeId": "9",
          "name": "@feature-tag1",
        },
        Object {
          "astNodeId": "10",
          "name": "@feature-tag2",
        },
        Object {
          "astNodeId": "5",
          "name": "@skip(browserName='chrome';platformName='windows')",
        },
        Object {
          "astNodeId": "6",
          "name": "@scenario-tag1",
        },
        Object {
          "astNodeId": "7",
          "name": "@scenario-tag2",
        },
      ],
      "title": undefined,
      "type": "step",
      "uid": "0",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events when executed in scenarioLeverReporter should send proper data on \`test-run-started\` event 1`] = `
Array [
  Array [
    "suite:start",
    Object {
      "cid": "0-1",
      "description": "  As a user of WebdriverIO
  I should be able to use different commands
  to get informations about elements on the page",
      "file": "/some/path/to/features/my-feature.feature",
      "keyword": "Feature",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "id": "9",
          "location": Object {
            "column": 1,
            "line": 1,
          },
          "name": "@feature-tag1",
        },
        Object {
          "id": "10",
          "location": Object {
            "column": 1,
            "line": 2,
          },
          "name": "@feature-tag2",
        },
      ],
      "title": "Example feature",
      "type": "feature",
      "uid": "my-feature.feature:undefined:1",
    },
  ],
]
`;

exports[`cucumber reporter emits messages for certain cucumber events when executed in scenarioLeverReporter should send proper data on onTestRunFinished 1`] = `
Array [
  Array [
    "suite:end",
    Object {
      "cid": "0-1",
      "duration": NaN,
      "file": "/some/path/to/features/my-feature.feature",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "id": "9",
          "location": Object {
            "column": 1,
            "line": 1,
          },
          "name": "@feature-tag1",
        },
        Object {
          "id": "10",
          "location": Object {
            "column": 1,
            "line": 2,
          },
          "name": "@feature-tag2",
        },
      ],
      "title": "Example feature",
      "type": "feature",
      "uid": "my-feature.feature:undefined:1",
    },
  ],
]
`;

exports[`cucumber reporter tags in title should add tags on handleBeforeFeatureEvent 1`] = `
Array [
  Array [
    "suite:start",
    Object {
      "cid": "0-1",
      "description": "  As a user of WebdriverIO
  I should be able to use different commands
  to get informations about elements on the page",
      "file": "/some/path/to/features/my-feature.feature",
      "keyword": "Feature",
      "specs": Array [
        "/foobar.js",
      ],
      "tags": Array [
        Object {
          "id": "9",
          "location": Object {
            "column": 1,
            "line": 1,
          },
          "name": "@feature-tag1",
        },
        Object {
          "id": "10",
          "location": Object {
            "column": 1,
            "line": 2,
          },
          "name": "@feature-tag2",
        },
      ],
      "title": "@feature-tag1, @feature-tag2: Example feature",
      "type": "feature",
      "uid": "my-feature.feature:undefined:1",
    },
  ],
]
`;
